### PR TITLE
Move optimize indexing to use the class Bounds

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102611
* __->__ #104558
* #104559
* #104557

This PR removes plenty of duplicated code. In particular, it removes the two repeated implementations of `get_expr_range`, which are superseded by the more correct `bound_sympy`.

The two duplicated `get_expr_range`s were a result of an oversight in https://github.com/pytorch/pytorch/pull/100549.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

### CODE REVIEW

**Comment by lezcano** [comment]
@eellison these three PRs are ready for review.

**Review by eellison — APPROVED** [review]
nice !!

File: torch/_inductor/bounds.py [hunk 1]
```diff
@@ -88,7 +19,8 @@ class BoundVars:
     def __init__(self, loop_body: LoopBody):
         self.loop_body = loop_body
         self.replacement_vals = {
-            k: ValueRanges(0, v) for k, v in loop_body.var_ranges.items()
+            k: ValueRanges(0, v if not free_symbols(v) else math.inf)
```
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/bounds.py:22]: Sort of. The thing is that here we are bounding operations of an FX graph, while in the other one we are bounding sympy expressions. The logic is pretty much the same though.

**Pull request closed without merging**